backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/GCSService.java

```diff
@@ -0,0 +1,183 @@
+          // If instead of returning the URL we received any other error message from the server for our post request,
+          // we can read it using getInputStream. Finally, we can thrown an error after reading the stream.
+          try (BufferedReader reader =
+              new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
+            String inputLine;
+            StringBuffer response = new StringBuffer();
+            while ((inputLine = reader.readLine()) != null) {
+              response.append(inputLine);
+            }
+
+            String errorMessage = response.toString();
+            LOGGER.error("Failed to send post request to GCS. Got the following error message: " + errorMessage);
+            throw new ModelDBException(errorMessage);
+          }
+        }
+      } finally {
+        connection.disconnect();
+      }
+    } catch (Exception ex) {
+      String errorMessage =
+          "Error while trying to send post request for getting resumable uploads signed URL";
+      LOGGER.error(errorMessage, ex);
+      throw new ModelDBException(errorMessage, ex);
+    }
+  }
+
+  private String generatePresignedUrl(String artifactPath, String method) throws ModelDBException {
+    // Validate bucket
+    Boolean exist = doesBucketExist(bucketName);
+    if (!exist) {
+      throw new ModelDBException("Bucket does not exists", io.grpc.Status.Code.INTERNAL);
+    }
+
+    LOGGER.trace("GCSService - generatePresignedUrl called");
+
+    // Creating blob object to pass information about bucket and path to upload/download artifacts
+    BlobInfo blobInfo = BlobInfo.newBuilder(BlobId.of(this.bucketName, artifactPath)).build();
+
+    if (method.equalsIgnoreCase(ModelDBConstants.PUT)) {
+      LOGGER.trace("GCSService - generatePresignedUrl - Put function called");
+      return getUploadUrl(blobInfo);
+    } else if (method.equalsIgnoreCase(ModelDBConstants.GET)) {
+      LOGGER.trace("GCSService - generatePresignedUrl - Get function called");
+      URL url =
+          this.gcsClient.signUrl(
+              blobInfo,
+              this.signedUrlValidity.getSeconds(),
+              TimeUnit.SECONDS,
+              Storage.SignUrlOption.withV4Signature(),
+              Storage.SignUrlOption.signWith((ServiceAccountSigner) this.credentials));
+
+      return url.toString();
+    } else {
+      String errorMessage = "Unsupported HTTP Method for GCS presigned URL";
+      Status status =
+          Status.newBuilder().setCode(Code.INVALID_ARGUMENT_VALUE).setMessage(errorMessage).build();
+      LOGGER.error(errorMessage);
+      throw StatusProto.toStatusRuntimeException(status);
+    }
+  }
+
+  @Override
+  public Optional<String> initiateMultipart(String s3Key) {
+    // Returning empty would let the underlying caller know that this artifact store method doesn't
+    // support multipart
+    // uploads and will only call generatePresignedUrl once to get the upload URL.
+    return Optional.empty();
```

Comment:
```suggestion
    String errorMessage = "Multipart not supported by GCS";
    Status status =
            Status.newBuilder().setCode(Code.FAILED_PRECONDITION_VALUE).setMessage(errorMessage).build();
    LOGGER.error(errorMessage);
    throw StatusProto.toStatusRuntimeException(status);
```
As per your code at https://github.com/VertaAI/modeldb/pull/2047/files#diff-cd7c39f9e673d88a84de50bef8afce24fdd1f0a8d293a477c25b88ad4b8937d2R181, we are not supporting the multipart on GCS so could we throw exception here to protect initialization of multipart on getUrlForArtifact call instead of commit multipart like i have suggested??

Reply:
@ad-47 : Thank you for the suggestion. I have accepted the commit suggestion.